File: tender-ai-backend/app/schemas/push.py
# -*- coding: utf-8 -*-
"""SL5 主動推播 schema（Layer B；對非合作範圍對象只回傳 Layer A 安全欄位）。

- 推播理由（reason）取自 reasoning.explain_tender 的 headline；分數（score）為
  criteria_fit（0–100）。皆為可解釋的聚合結果，**不含人名／email 或個別評語原文**。
- 每筆推播卡同時帶該標案的 Layer A 顯示欄位（名稱／機關／類別／城市／預算／
  ROC 截止／剩餘天數／來源／連結），供前端通知面板直接呈現。
"""
from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field


class PushRunRequest(BaseModel):
    """觸發一次每日推播批次（手動或排程）。"""

    user_id: int | None = None
    # 單次最多推幾筆（高潛力上限）
    limit: int = Field(default=8, ge=1, le=50)
    # 顯示可行度（feasibility_score）門檻：≥ 才納入候選
    min_score: int = Field(default=60, ge=0, le=100)
    # 跨日去重視窗：近 N 天已推過的標案不重複推
    lookback_days: int = Field(default=7, ge=0, le=90)


class PushItemOut(BaseModel):
    """單筆推播卡（push_log + 該標案 Layer A 顯示欄位）。"""

    model_config = ConfigDict(from_attributes=True)

    id: int  # push_log id
    tender_id: int | None
    run_date: date
    score: int | None
    tier: str | None
    reason: str | None
    channel: str
    status: str
    pushed_at: datetime
    read_at: datetime | None
    # ---- Layer A 標案顯示欄位（tender 被刪時為 None） ----
    name: str | None = None
    org: str | None = None
    category: str | None = None
    city: str | None = None
    budget_wan: int | None = None
    deadline_roc: str | None = None
    days_left: int | None = None
    source: str | None = None
    link: str | None = None


class PushRunResult(BaseModel):
    """一次批次結果：新建幾筆、因去重略過幾筆，與該批次的推播卡。"""

    run_date: date
    created: int
    skipped: int
    items: list[PushItemOut]


class PushDigestOut(BaseModel):
    """通知面板資料：最新一批推播卡 + 跨全部批次的未讀數。"""

    run_date: date | None = None
    unread: int = 0
    total: int = 0
    items: list[PushItemOut] = Field(default_factory=list)


class PushReadRequest(BaseModel):
    """標記已讀：push_id 給定 → 單筆；省略 → 該使用者全部未讀。"""

    user_id: int | None = None
    push_id: int | None = None


class PushReadResult(BaseModel):
    marked: int

File: tender-ai-frontend/src/lib/push.ts
// SL5 主動推播 client：對接後端 push API（皆 Layer A 安全內容、user_id 隔離）。
//   GET  /push/digest → 通知面板資料（最新一批推播卡 + 跨批次未讀數）
//   POST /push/run    → 觸發一次每日推播批次（依承標判準挑高潛力標案）
//   POST /push/read   → 標記已讀（單筆或全部）
// 契約見 tender-ai-backend/app/schemas/push.py。沿用 assistant.ts 的 API_BASE / authHeaders。
import { API_BASE } from "@/lib/api-base";
import type { Tier } from "@/types/domain";

function authHeaders(): Record<string, string> {
  const key = import.meta.env.VITE_API_KEY as string | undefined;
  return key ? { "X-API-Key": key } : {};
}

// 對齊後端 PushItemOut（snake_case → camelCase）。皆為標案公開（Layer A）欄位
// 與可解釋聚合分數／理由，不含人名／email。
export interface PushItem {
  id: number;
  tenderId: number | null;
  runDate: string;
  score: number | null;
  tier: Tier | null;
  reason: string | null;
  channel: string;
  status: string;
  pushedAt: string | null;
  readAt: string | null;
  // Layer A 顯示欄位
  name: string | null;
  org: string | null;
  category: string | null;
  city: string | null;
  budgetWan: number | null;
  deadlineRoc: string | null;
  daysLeft: number | null;
  source: string | null;
  link: string | null;
}

export interface PushDigest {
  runDate: string | null;
  unread: number;
  total: number;
  items: PushItem[];
}

export interface PushRunResult extends PushDigest {
  created: number;
  skipped: number;
}

interface PushItemRaw {
  id: number;
  tender_id: number | null;
  run_date: string;
  score: number | null;
  tier: Tier | null;
  reason: string | null;
  channel: string;
  status: string;
  pushed_at: string | null;
  read_at: string | null;
  name: string | null;
  org: string | null;
  category: string | null;
  city: string | null;
  budget_wan: number | null;
  deadline_roc: string | null;
  days_left: number | null;
  source: string | null;
  link: string | null;
}

function adaptItem(r: PushItemRaw): PushItem {
  return {
    id: r.id,
    tenderId: r.tender_id,
    runDate: r.run_date,
    score: r.score,
    tier: r.tier,
    reason: r.reason,
    channel: r.channel,
    status: r.status,
    pushedAt: r.pushed_at,
    readAt: r.read_at,
    name: r.name,
    org: r.org,
    category: r.category,
    city: r.city,
    budgetWan: r.budget_wan,
    deadlineRoc: r.deadline_roc,
    daysLeft: r.days_left,
    source: r.source,
    link: r.link,
  };
}

/** 取得通知面板資料。後端未啟動／錯誤時 throw，由 UI fallback。 */
export async function fetchPushDigest(): Promise<PushDigest> {
  const res = await fetch(`${API_BASE}/push/digest`, {
    headers: { ...authHeaders() },
  });
  if (!res.ok) throw new Error(`push digest ${res.status}`);
  const d = (await res.json()) as {
    run_date: string | null;
    unread: number;
    total: number;
    items: PushItemRaw[];
  };
  return {
    runDate: d.run_date,
    unread: d.unread,
    total: d.total,
    items: (d.items ?? []).map(adaptItem),
  };
}

/** 觸發一次每日推播批次（手動產生）。同日重跑 idempotent。 */
export async function runPush(opts?: {
  limit?: number;
  minScore?: number;
  lookbackDays?: number;
}): Promise<PushRunResult> {
  const body: Record<string, number> = {};
  if (opts?.limit != null) body.limit = opts.limit;
  if (opts?.minScore != null) body.min_score = opts.minScore;
  if (opts?.lookbackDays != null) body.lookback_days = opts.lookbackDays;

  const res = await fetch(`${API_BASE}/push/run`, {
    method: "POST",
    headers: { "Content-Type": "application/json", ...authHeaders() },
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(`push run ${res.status}`);
  const d = (await res.json()) as {
    run_date: string | null;
    created: number;
    skipped: number;
    items: PushItemRaw[];
  };
  return {
    runDate: d.run_date,
    created: d.created,
    skipped: d.skipped,
    unread: d.items.filter((i) => i.status === "pending").length,
    total: d.items.length,
    items: d.items.map(adaptItem),
  };
}

/** 標記已讀。pushId 給定 → 單筆；省略 → 全部未讀。回傳更新筆數。 */
export async function markPushRead(pushId?: number): Promise<number> {
  const res = await fetch(`${API_BASE}/push/read`, {
    method: "POST",
    headers: { "Content-Type": "application/json", ...authHeaders() },
    body: JSON.stringify(pushId != null ? { push_id: pushId } : {}),
  });
  if (!res.ok) throw new Error(`push read ${res.status}`);
  const d = (await res.json()) as { marked: number };
  return d.marked;
}
